### ISSUE TITLE
[flang][Semantics] Semantic error in Fortran code due to recursive reference in specification expression

### ISSUE THREAD

**Issue by inaki-amatria** [llvm/llvm-project#106705]
- `flang-new` version is: `flang-new version 20.0.0git (https://github.com/llvm/llvm-project.git 46fe36a4295f05d5d3731762e31fc4e6e99863e9)`
- Reduced test case is:
```fortran
! foo.f90

module foo
  real :: qux(10, 10, 10)

  interface
    module subroutine bar(i)
    end
    module function baz()
    end
  end interface
end

submodule (foo) foo_impl
  contains
    module procedure bar
      qux(i, i, i) = baz()
    end
    module procedure baz
    end
end
```
- `flang-new` invocation is: `flang-new foo.f90`
- Expected behavior is: The code compiles sucessfully without errors, as `baz` should be accessible through host association in `bar`.
- Actual behavior: Compilation fails with the following error:
```txt
error: Semantic errors in /foo.f90
foo.f90:17:22: error: The module function 'baz' may not be referenced recursively in a specification expression
        qux(i, i, i) = baz()
                       ^^^^^
```
- Additional information: Reordering the definitions of the procedures in the submodule seems to work around the issue for now.

**Comment by klausler** [comment]
Semantics has to sort out the syntactically ambiguous statement `qux(i,i,i)=baz()` statement, which can be either a statement function definition or an assignment statement, but hadn't yet done so by the time some other analysis was performed, leading to a very confusing error message; sorry about that.  If you need a temporary work-around, put a `continue` statement (or any other executable part construct) before that statement.  Will fix.

**Comment by klausler** [comment]
Fix coming in https://github.com/llvm/llvm-project/pull/106768.

**Comment by inaki-amatria** [comment]
> Fix coming in #106768.

Wow! That was quick. Thank you so much @klausler !
